lib/navigation/preprocess_imu.cpp

```diff
@@ -9,48 +14,126 @@ ImuPreprocessor::ImuPreprocessor()
+  checkReliable();
+  return clean_accelerometer_data;
 }
 
 core::ImuData ImuPreprocessor::detectOutliers(const core::RawImuData imu_data)
 {
-  /*
-  TODOLater: implement
-  rough process:
-  - get q1, median, q3 of encoder array
-  - define upper & lower bounds as (-)1.5*inter-quatrile range
-  - if any datapoint outwith this, set
-  outlier_imus_[i] += 1
-  -set outlier points as median
-
-  -also has to be able to handle 1 unreliable sensor
-  -also figure out return type/ what we update and update
-  documentation as appropriate
-  */
-  return {0, 0, 0, 0};
+  core::ImuData clean_accelerometer_data;
+  // TODOLater : change magnitude for directions (hardware configuration?)
+  core::Float magnitude;
+  for (std::size_t i = 0; i < core::kNumImus; ++i) {
+    magnitude = 0;
+    for (std::size_t j = 0; j < 3; ++j) {
+      magnitude += std::pow(imu_data.at(i).at(j), 2);
+    }
+    clean_accelerometer_data.at(i) = std::sqrt(magnitude);
+  }
+  Quartiles quartiles     = getOutlierThresholds(clean_accelerometer_data);
+  core::Float iqr         = quartiles.q3 - quartiles.q1;
+  core::Float lower_bound = quartiles.median - 1.5 * iqr;
+  core::Float upper_bound = quartiles.median + 1.5 * iqr;
+  for (size_t i = 0; i < core::kNumImus; ++i) {
+    // converts outliers or unreliables to medians, updates number of consecutive outliers for each
+    // sensor
+    if (are_imus_reliable_.at(i) == false) {
+      clean_accelerometer_data.at(i) = quartiles.median;
+    } else if (clean_accelerometer_data.at(i) < lower_bound
+               || clean_accelerometer_data.at(i) > upper_bound) {
+      clean_accelerometer_data.at(i) = quartiles.median;
+      ++num_outliers_per_imu_.at(i);
+    } else {
+      num_outliers_per_imu_.at(i) = 0;
+    }
+  }
+  return clean_accelerometer_data;
+}
+
+Quartiles ImuPreprocessor::getOutlierThresholds(const core::ImuData &clean_accelerometer_data)
+{
+  core::ImuData clean_accelerometer_data_copy_all_reliable;
+  std::array<core::Float, core::kNumImus - 1> clean_accelerometer_data_copy_one_unreliable;
+  const std::size_t num_reliable_accelerometers
+    = std::accumulate(are_imus_reliable_.begin(), are_imus_reliable_.end(), 0);
+  if (num_reliable_accelerometers == core::kNumImus) {
+    std::copy(clean_accelerometer_data.begin(),
+              clean_accelerometer_data.end(),
+              clean_accelerometer_data_copy_all_reliable.begin());
+    std::sort(clean_accelerometer_data_copy_all_reliable.begin(),
+              clean_accelerometer_data_copy_all_reliable.end());
+  } else {
+    std::size_t index;
+    for (size_t i = 0; i < core::kNumImus - 1; ++i) {
+      if (!are_imus_reliable_.at(i)) { index = i; }
+    }
+    for (size_t i = 0; i < core::kNumImus - 1; ++i) {
+      if (i >= index) {
+        clean_accelerometer_data_copy_one_unreliable.at(i) = clean_accelerometer_data.at(i + 1);
+      } else {
+        clean_accelerometer_data_copy_one_unreliable.at(i) = clean_accelerometer_data.at(i);
```

Comment:
`clean_accelerometer_data_copy_one_unreliable`?!?!?!?!?! I want to rip my eyes out!

Reply:
If you start using names like that, you're making too many variables!

Reply:
yeah a very bad place holder name. I'm not too sure how to cut down on the amount of variables, as I need to initialise both arrays for different situations, as I can't initialise in loops without massively repetitive code

Reply:
Templates and helper functions where possible. Otherwise just more careful naming.